receiver/m365receiver/client.go

```diff
@@ -0,0 +1,441 @@
+	}
+
+	defer func() { _ = resp.Body.Close() }()
+
+	// troubleshoot resp err if present
+	if resp.StatusCode != 200 {
+		// get error code
+		respErr := struct {
+			Err string `json:"error"`
+		}{}
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return err
+		}
+		err = json.Unmarshal(body, &respErr)
+		if err != nil {
+			return err
+		}
+		//match on error code
+		switch respErr.Err {
+		case "unauthorized_client":
+			return fmt.Errorf("the provided client_id is incorrect or does not exist within the given tenant directory")
+		case "invalid_client":
+			return fmt.Errorf("the provided client_secret is incorrect or does not belong to the given client_id")
+		case "invalid_request":
+			return fmt.Errorf("the provided tenant_id is incorrect or does not exist")
+		}
+		return fmt.Errorf("got non 200 status code from request, got %d", resp.StatusCode)
+	}
+
+	// read in access token
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+	token := struct {
+		Token string `json:"access_token"`
+	}{}
+	err = json.Unmarshal(body, &token)
+	if err != nil {
+		return err
+	}
+	m.token = token.Token
+	return nil
+}
+
+// function for getting metrics data
+func (m *m365Client) GetCSV(ctx context.Context, endpoint string) ([]string, error) {
+	resp, err := m.makeRequest(ctx, nil, "GET", endpoint, WithToken(m.token))
+	if err != nil {
+		return []string{}, err
+	}
+
+	//troubleshoot resp err if present
+	if resp.StatusCode != 200 {
+		return []string{}, m.handleErrors(resp)
+	}
+
+	defer func() { _ = resp.Body.Close() }()
+	csvReader := csv.NewReader(resp.Body)
+
+	//parse out 2nd line & return csv data
+	_, err = csvReader.Read()
+	if err != nil {
+		return []string{}, err
+	}
+	data, err := csvReader.Read()
+	if err != nil {
+		if err == io.EOF { //no data in report, scraper should still run
+			return []string{}, nil
+		}
+		return []string{}, err
+	}
+
+	return data, nil
+}
+
+// function for getting log data
+func (m *m365Client) GetJSON(ctx context.Context, endpoint string, end string, start string) ([]logData, error) {
+	// make request to Audit endpoint
+	resp, err := m.makeRequest(ctx, nil, "GET", endpoint, WithToken(m.token), WithTime(end, start))
+	if err != nil {
+		return []logData{}, err
+	}
+	defer func() { _ = resp.Body.Close() }()
+
+	// troubleshoot error code
+	if resp.StatusCode != 200 {
+		return []logData{}, m.handleErrors(resp)
+	}
+
+	// get redirect links
+	redirectLinks, err := m.readInContent(resp)
+	if err != nil {
+		return []logData{}, err
+	}
+	if redirectLinks == nil { // if no data, return no err
+		return []logData{}, nil
+	}
+
+	// follows redirect link(s) to get json array of logs
+	// parse json array into individual strings, still in json
+	// append to data each body and corresponding json struct
+	var data = []logData{}
+	for _, l := range redirectLinks {
+		body, err := m.followLink(ctx, l)
+		if err != nil {
+			return []logData{}, err
+		}
+		bodies := m.parseBody(body)
```

Comment:
I see what you're doing here with parseBody, but I think there's a better approach here than doing string manipulations like this, by unmarshalling into a slice of json.RawMessage first.

Here's a short example of what I mean:
https://go.dev/play/p/0L01vLDlLTs?v=goprev